modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -66,6 +66,43 @@ def __init__(self, data, length=None, width=None, ip=None, call_queue=None):
             )
         )
 
+    @staticmethod
```

Comment:
I want us to keep the code for Ray engine in sync with unidist. If we enable unidist on Ray, then we will have an issue similar to one this PR resolves. So please add the respective changes to unidist engine. Perf measurements would also be useful.

Reply:
well, I can't test this with unidist because of modin-project/unidist#354

Reply:
The issue was fixed on unidist side and the fix will be available in unidist 0.5.0. I suppose you could test the changes without .wait method.

Reply:
> We could also release unidist 0.4.2 with that fix if it is really necessary.

I think I'll just test with the master branch of unidist, so no urgent for the release

Reply:
Fine. Please take into account the enabling or disabling of the shared object store (UNIDIST_MPI_SHARED_OBJECT_STORE). By default, it is disabled.

Reply:
Well, I tried [the reproducer](https://github.com/modin-project/modin/issues/4507#issuecomment-1748641449) from the issue, and it seems to work fine with unidist on MPI with both shared memory disabled and enabled. Since the change in the PR is not a feature but a bug fix, do you think it makes sense to port them to the unidist backend even if the problem is not reproducible with unidist? I would prefer not to touch something if it's not broken.

Reply:
> I want us to keep the code for Ray engine in sync with unidist

I've just looked at the code of [unidist partitioning](https://github.com/modin-project/modin/blob/master/modin/core/execution/unidist/implementations/pandas_on_unidist/partitioning/partition.py), and it looks 100% identical to the [Ray's partitions code](https://github.com/modin-project/modin/blob/master/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py). Maybe it's better to come-up with some unified classes for both ray and unidist instead of copying the code?

Reply:
> do you think it makes sense to port them to the unidist backend even if the problem is not reproducible with unidist?

I think it makes sense to port the changes to the unidist backend because as I said if we enable unidist on Ray in future, then we will have an issue similar to one this PR resolves.

> I've just looked at the code of [unidist partitioning](https://github.com/modin-project/modin/blob/master/modin/core/execution/unidist/implementations/pandas_on_unidist/partitioning/partition.py), and it looks 100% identical to the [Ray's partitions code](https://github.com/modin-project/modin/blob/master/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py). Maybe it's better to come-up with some unified classes for both ray and unidist instead of copying the code?

We already had a similar discussion when integrating unidist into Modin. We ported changes from Ray engine to unidist and just replaced ray.* call to unidist.* calls. If we came up with some unified classes for both ray and unidist instead of copying the code, I assume we would have criss-crossed dependencies for Ray and unidist, which we don't want. Since unidist has Ray-like API, it looks like code duplication. If unidist had different API from Ray, we wouldn't see code duplication.

Reply:
I tried to port the changes to unidist, but it doesn't work very well
![image](https://github.com/modin-project/modin/assets/62142979/4437c696-b5cf-4f12-b1d2-aabf90099e77)

Reply:
I also tried to compare these two approaches with pure unidist and have been able to reproduce the deadlock caused by calling `unidist.get()` inside a kernel... Strange that it didn't happen with modin
```python
import unidist
import pandas as pd
from timeit import default_timer as timer
import warnings

unidist.init()

def gen_df():
    return pd.DataFrame(range(10))

@unidist.remote
def func1(*args):
    return args

@unidist.remote
def func2(args):
    res = []
    for a in args:
        res.append(unidist.get(a))
    return res

remote_objs = [unidist.put(gen_df()) for _ in range(3)]

warnings.warn("running func1...")
t1 = timer()
unidist.wait(func1.remote(*remote_objs)) # works fine
print(timer() - t1)

remote_objs = [unidist.put(gen_df()) for _ in range(3)]

warnings.warn("running func2...")
t1 = timer() 
unidist.wait(func2.remote(remote_objs)) # deadlock
print(timer() - t1)
```

Reply:
> Strange that it didn't happen with modin

unidist on MPI internally deconstructs all nested object refs and pass those into a kernel so data is materialized in the kernel. However, your example looks as a problem in unidist anyway. Can you file an issue in unidist. We will take a look at it.